### ISSUE TITLE
Add an enabledProperty to ABSwitch

### ISSUE THREAD

**Issue by arouinfar** [phetsims/sun#571]
In https://github.com/phetsims/ph-scale/issues/117 @kathy-phet and I reviewed the ABSwitch in depth in studio. The changes made in #559 are looking great in studio, but the one thing that we think is missing is a top-level enabledProperty. Currently a client could make an ABSwitch look disabled by adjusting opacityProperty and pickableProperty, but that's a bit clunky. 

In today's status meeting @pixelzoom estimated an additional four hours to add the enabledProperty, and reminded us that there is a bigger-picture discussion that we should have about enabledProperty, in general. @kathy-phet said that adding an enabledProperty to ABSwitch is not a high priority issue right now.

Assigning to the PhET-iO team.

**Comment by pixelzoom** [comment]
ToggleSwitch is now the base class of this UI component hierachy, so that's where this would be added.  I'll change the title of this issue to reflect that. 

We can certainly add enabledProperty to ToggleSwitch.   But PhET's current way of implementing enabledProperty is to add it to each UI component, essentially duplicating the same code/pattern over and over.  A longterm solution would be preferable, so please consider putting resources into https://github.com/phetsims/sun/issues/257.

**Comment by zepumph** [comment]
I added this by creating a UIComponent mixin.  see https://github.com/phetsims/sun/issues/257 for details. ToggleSwitch is currently the only usage of it, and this issue should be out for review with https://github.com/phetsims/sun/issues/257 (review of the whole mixin).

**Comment by zepumph** [comment]
This isn't quite complete. Though there is a working enabledProperty in ToggleSwitch, we need to add a listener in ABSwitch to toggle pickability and opacity.

**Comment by zepumph** [comment]
In the above commit, I added the listener to ABSwitch so that it can look disabled when the ToggleSwitch is disabled. @jbphet does this look good to you?

**Comment by pixelzoom** [comment]
I suspect a complaint with https://github.com/phetsims/sun/commit/62d4c17c4165f60012d6d9dec8202c262085f60e is that may be that designers will want an `enabledProperty`  associated with the top-level UI component, e.g.:

```
-myABSwitch
  enabledProperty
  +labelANode
  +labelBNode
  +toggleSwitch
```

In the current implementation, you're going to have something like:

```
-myABSwitch
  +labelANode
  +labelBNode
  -toggleSwitch
    enabledProperty
```

But I think that this is an inherent problem with composition (as used in ABSwitch), and we need to may need to live with this.   The alternative would be for ToggleSwitch and ABSwitch to both have `enabledProperty`, like this:

```
-myABSwitch
  enabledProperty
  +labelANode
  +labelBNode
  +toggleSwitch
    enabledProperty
```

And that could lead to all kinds of problems. For example, setting both `myABSwitch. enabledProperty.value = true` and `myABSwitch.toggleSwitch.enabledProperty.value = true` might result in the ToggleSwitch appearing to be doubly enabled (i.e. the opacity reduced twice - one by ToggleSwitch, once by ABSwitch).

**Comment by pixelzoom** [comment]
Before working on this (and before doing any more work on this), I would ask the designer how they want this to look in the scenery tree.

**Comment by arouinfar** [comment]
@pixelzoom's first example in https://github.com/phetsims/sun/issues/571#issuecomment-592714040 was what I meant by "top-level enabledProperty" to ABSwitch:

```
-myABSwitch
  enabledProperty
  +labelANode
  +labelBNode
  +toggleSwitch
```

I didn't realize the work would need to be done in ToggleSwitch, but @pixelzoom's https://github.com/phetsims/sun/issues/571#issuecomment-592281295 clarified that.

The current implementation in master with `myABSwitch.toggleSwitch.enabledProperty` has the desired behavior. For example, setting ` phScale.microScreen.view.graphNode.graphScaleSwitch.toggleSwitch.enabledProperty` to false results in the entire ABSwitch being disabled:
![image](https://user-images.githubusercontent.com/8419308/75588613-43c5ed00-5a36-11ea-9d0f-5ba68688d4f8.png)

I do not think the current tree structure would be the most intuitive to instructional designers, but I understand why it is the way it is. I would rather leave it as-is than pursue @pixelzoom's 3rd suggestion in https://github.com/phetsims/sun/issues/571#issuecomment-592714040.

**Comment by pixelzoom** [comment]
@zepumph and I discussed this on Zoom and made some refinements:

`ToggleSwitch` looks good, it has its own `enabledProperty`, which is `phetioFeatured: true`.  The mixin `UIComponent` looks nice, but will be renamed to `EnabledComponent`, which better reflects its responsibilities. 

ABSwitch composes 3 Nodes: 2 labels (provided by the caller) and a ToggleSwitch.  But ABSwitch  needs to have its own `enabledProperty`, not reach into ToggleSwitch. This is similar to NumberControl, where we don't reach into Slider and use its enabledProperty. 

ABSwitch will set its ToggleSwitch's enabledProperty to `phetioFeatured: false` so that it isn't features.  But like a Slider in a NumberControl, someone can still reach into an ABSwitch and change `enabledProperty` for its ToggleSwitch.

**Comment by pixelzoom** [comment]
So to summarize, here's what you'll see for PhET-iO Featured:

For a ToggleSwitch:

```
-myToggleSwitch
  enabledProperty
  visibleProperty
```

For an OnOffSwitch (which is a subclass of ToggleSwitch):

```
-myOnOffSwitch
  enabledProperty
  visibleProperty
```

For an ABSwitch (which uses a ToggleSwitch via composition):

```
-myABSwitch
  enabledProperty
  visibleProperty
  +labelANode
  +labelBNode
  -toggleSwitch
    visibleProperty
```

**Comment by zepumph** [comment]
Agreed and thanks for the summary @pixelzoom. I will rename to EnabledComponent over in #257 for clarity in that review. 

Everything else is implemented above except one thing:

>For an ABSwitch (which uses a ToggleSwitch via composition):

>>-myABSwitch
  enabledProperty
  visibleProperty
  **+labelANode
  +labelBNode**
  -toggleSwitch
    visibleProperty


Note that as explained and agreed upon in https://github.com/phetsims/sun/issues/572#issuecomment-592724144, the labels are injected dependencies, and so that phetioFeatured implementation is not added to code in ABSwitch, and will either be done in overrides (most likely) or in usages if better.

@arouinfar please review the enabledProperty implementation in Studio, and I'll leave this assigned to @jbphet so that he can see two usages of the new mixin. It is now used in ABSwitch and ToggleSwitch.

**Comment by pixelzoom** [comment]
Thanks for the clarification on the label nodes.  Agreed.

**Comment by arouinfar** [comment]
Things are looking great in studio, thanks @pixelzoom @zepumph!

Here's an example of the studio tree for the graphScaleSwitch in pH Scale
![image](https://user-images.githubusercontent.com/8419308/75594519-3a448100-5a46-11ea-9c98-06e84effb01d.png)

When viewing the featured elements, the tree is also looking good (A & B elements will be featuerd in overrides, as decided in #572)
![image](https://user-images.githubusercontent.com/8419308/75594637-9d361800-5a46-11ea-891c-aa5d1a25ce62.png)

**Comment by kathy-phet** [comment]
@arouinfar - I'm confused as to why the lower image has enabledProperty at the top level, but the upper image has it in the toggleSwitch? Did the common code change between these images?

**Comment by arouinfar** [comment]
@kathy-phet it's because I had two versions of ph-scale open, and I took the screenshot from the wrong tab. I've updated the screenshot.

**Comment by pixelzoom** [comment]
@zepumph It looks the name change to `EnabledComponent` was not applied to methods `disposeUIComponent` and `_disposeUIComponent`.   Fixed in the above commit, please review.

**Comment by jbphet** [comment]
Two review comments:

+ The call to `this.disposeEnabledComponent()` seems to me like it should be in the `dispose` function, not the internal `this.disposeToggleSwitch`.  My reasoning here is that this is much like calling `super.dispose()`.
+ There is a header comment above the class declaration that says "@mixes {EnabledComponent}".  This seems like a good idea, since it will allow us to look in one place to see all of the things that the class extends or mixes.  Can this be enforced with a lint rule?

**Comment by zepumph** [comment]
>@zepumph It looks the name change to EnabledComponent was not applied to methods disposeUIComponent and _disposeUIComponent. Fixed in the above commit, please review.

Excellent. Thanks. I also hit the initialize method. 

>The call to this.disposeEnabledComponent() seems to me like it should be in the dispose function, not the internal this.disposeToggleSwitch. My reasoning here is that this is much like calling super.dispose().

Great idea, I implemented that, and did the same for ABSwitch.

>There is a header comment above the class declaration that says "@mixes {EnabledComponent}". This seems like a good idea, since it will allow us to look in one place to see all of the things that the class extends or mixes. Can this be enforced with a lint rule?

I can't link of an easy way to make a lint rule for this. Perhaps an item in the CRC like "any traits or mixins for a class should be annotated with `@mixes MyMixin`. I don't feel too strongly, as I think this is largely done already (46 usages of `@mixes MyType`). I added an item to the CRC. 


@jbphet do you think this is ready to close?
